client/rhel/spacewalk-client-tools/man/en/up2date.5
.\" Copyright 1999-2010 Red Hat, Inc.
.\"
.\" This man page is free documentation; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; version 2 of the License.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this man page; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH "UP2DATE" "5" "2010 August 30" "Linux" "Red Hat, Inc."
.SH NAME

up2date \- Configuration file for Spacewalk client programs.

.SH DESCRIPTION

.PP
Spacewalk client programs, such as \fBrhn_check(8)\fR, use a common
configuration file at \fI/etc/sysconfig/rhn/up2date\fR.

.SH OPTIONS

.IP \fBenableProxy\fR
Use a HTTP Proxy. Can be either 0 (False) or 1 (True).
.br
.IP \fBserverURL\fR
Remote server URL
.br
.IP \fBdebug\fR
Whether or not debugging is enabled. Can be either 0 (False) or 1 (True).
.br
.IP \fBsystemIdPath\fR
Location of system id file. Expects the path of a file on disk.
.br
.IP \fBversionOverride\fR
Override the automatically determined system version. Expects a string or blank (do not override the version).
.br
.IP \fBhttpProxy\fR
HTTP proxy in host:port format, e.g. squid.redhat.com:3128
.br
.IP \fBproxyUser\fR 
The username for an authenticated proxy.
.br
.IP \fBproxyPassword\fR
The password to use for an authenticated proxy.
.br
.IP \fBenableProxyAuth\fR
To use an authenticated proxy or not. Can be either 0 (False) or 1 (True).
.br
.IP \fBnetworkRetries\fR 
Number of attempts to make at network connections before giving up. Expects a
positive integer.
.br
.IP \fBsslCACert\fR
The CA cert used to verify the ssl server. Expects the path of a file on disk.
.br
.IP \fBnoReboot\fR
Disable the reboot action. Can be either 0 (False) or 1 (True).
.br
.IP \fBdisallowConfChanges\fR
Config options that can not be overwritten by a config update action. Expects a
semi-colon separated list of config option names.
.IP \fBhostedWhitelist\fR
Override the default list of URLs the RHN Register client treats as official RHN hosted servers. This only affects the RHN Register graphical- and text-based clients, as it changes some of the options that appear. This option is included for testing purposes only.
.br
.IP \fBretrieveOnly\fR
If set to 1 (True), package installs done through rhn_check will only download
the package, the package will not be installed.
.br
.SH FILES

.IP \fI/etc/sysconfig/rhn/up2date\fP
The location of this configuration file

.SH "SEE ALSO"

.PP
\fBrhn_check\fP(8), \fBrhnreg_ks\fP(8), \fBrhnplugin\fP(8), \fBrhnplugin.conf\fP(5).

.SH AUTHORS
.PP
See the AUTHORS file included with this software.
.PP
This manual page was written by James Bowes <jbowes@redhat.com>

.SH "BUGS"
.PP
Report bugs to <http://bugzilla.redhat.com>.

.SH COPYRIGHT

.PP
Copyright \(co 1999\-2010 Red Hat, Inc.

.PP
This is free software; see the source for copying conditions.  There is 
NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


docs/changes.md
# What is different between the master suite and the slenkins suite

#### What has changed? (quick summary for impatients)
  
  1) Twopence -> replaced almostly the sshcmd call.
  1b) Lavanda (ruby library) for expanding function over targets:
  https://github.com/MalloZup/Lavanda
  
  2) packages are not in kiwi images, but are installed dinamically by run.py
  
  3) nothing hardcoded on specific images. 


Basically the slenkins-suite use twopence, this allow to define targets for remote command execution.

In the SLEnkins suite, is added in Gemfile the requirement Twopence, and in the support directory,
twopence.init.rb
https://github.com/SUSE/spacewalk-testsuite-base/blob/slenkins/features/support/twopence_init.rb#L4

```
$server_ip = ENV['TESTHOST']
$client_ip = ENV['CLIENT']

$client = Twopence::init("ssh:#{$client_ip}")
$server = Twopence::init("ssh:#{$server_ip}")
```

In this snippet, we define 2 target, that use ssh as protocol for communication. (virtio is another protocl, but for cloud is not already supported).

The machines(vms), network, disk-space, ram ares setted up dinamically by slenkins-engine, so nothing is hardcoded. (and this parameters can be setted on the node file)
https://github.com/SUSE/spacewalk-testsuite-base/blob/slenkins/slenkins/tests-control/nodes


Some changes in testcases of cucumber were added, because in SLEnkins suite the cucumber features and step_definitions resides all in the control_node.
The control-node is a systemd-container. SLE-12-sp1 that execute the suite on system under tests SUT. (these are twopence targets)

At moment, we have 3 machines: server, client (is without salt), minion (salt client).

For a quick design explanation, look here:
https://github.com/okirch/susetest/blob/master/doc/susetest_design.jpg


An example for a change is here: **features/step_definitions/register_client_steps.rb**

Master
```
Given(/^I am root$/) do
  uid = `id -u`
  if ! $?.success? || uid.to_i != 0
    raise "You are not root!"
  end
  if $myhostname == "linux"
    raise "Invalid hostname"
  end
end
```
SLEnkins
```
Given(/^I am root$/) do
  user, local, remote, code = $client.test_and_store_results_together("whoami", "root", 500)
  if  user.strip != "root"
    puts  "user on client was #{user}" 
    raise "You are not root!"
  end
end
```

So the main difference, that is overall on the features/step_definitions, is :
1) Twopence commands, that give a target -->
This imply No more commands without Targets specification. (it make also the code more readeable, and ordered, since we know where is executed (on server, client, minion)


1) ```user, local, remote, code = $client.test_and_store_results_together("whoami", "root", 500)``

user is th output.

code is the retcode from command.

the other 2 vars, are for debugging twopence and for testing we can avoid to use them.

You can specify easyly target ( this are defined by twopence init in support)
```user, local, remote, code = $server.test_and_store_results_together("whoami", "root", 500)``

Root is the user that will execute the cmd. You can specify others.
500 is the timeout

In other tests, we have like this:

```
out, local, remote, code = $client.test_and_store_results_together("yum clean all", "root", "600")
fail if code != 0
```








python/spacewalk/server/handlers/xmlrpc/up2date.py
#  pylint: disable=missing-module-docstring
#
# Copyright (c) 2008--2016 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
#
# Red Hat trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate Red Hat trademarks that are incorporated
# in this software or its documentation.
#
# Implements the up2date.* fucntions for XMLRPC
#

# system module import
import time

from spacewalk.server.rhnServer import server_lib
from uyuni.common.usix import ListType, TupleType, StringType

# pylint: disable-next=ungrouped-imports
from spacewalk.common import rhnFlags
from spacewalk.common.rhnLog import log_debug, log_error
from spacewalk.common.rhnConfig import CFG
from spacewalk.common.rhnException import rhnFault
from spacewalk.common.rhnTB import add_to_seclist
from spacewalk.common.rhnTranslate import _
from spacewalk.server.rhnLib import computeSignature
from spacewalk.server.rhnHandler import rhnHandler
from spacewalk.server import rhnChannel, rhnDependency, rhnCapability
from spacewalk.server.rhnServer import server_route

import re

NONSUBSCRIBABLE_CHANNELS = re.compile("(rhn-proxy|rhn-satellite)")


class Up2date(rhnHandler):

    """xml-rpc Server Functions that we will provide for the outside world."""

    def __init__(self):
        """Up2date Class Constructor

        o Initializes inherited class.
        o Appends the functions available to the outside world in the
          rhnHandler list.
        """
        rhnHandler.__init__(self)
        # Function list inherited from rhnHandler
        # This action garners control of what is available to the client.

        # --- Clients v2+ ---
        # (getting headers, source and packages done with GETs now).
        self.functions.append("login")
        self.functions.append("listChannels")
        self.functions.append("subscribeChannels")
        self.functions.append("unsubscribeChannels")
        self.functions.append("history")
        self.functions.append("solvedep")
        self.functions.append("solveDependencies")
        self.functions.append("solveDependencies_arch")
        self.functions.append("solveDependencies_with_limits")

    def auth_system(self, action, system_id):
        # Stuff the action in the headers:
        transport = rhnFlags.get("outputTransportOptions")
        transport["X-RHN-Action"] = action
        return rhnHandler.auth_system(self, system_id)

    # pylint: disable-next=dangerous-default-value,unused-argument
    def login(self, system_id, extra_data={}):
        """Clients v2+
        Log in routine.
        Return a dictionary of session token/channel information.
        Also sets this information in the headers.
        """
        log_debug(5, system_id)
        # Authenticate the system certificate. We need the user record
        # to generate the tokens
        self.load_user = 1
        # pylint: disable-next=unused-variable
        server = self.auth_system("login", system_id)
        # log the entry
        log_debug(1, self.server_id)
        # Update the capabilities list
        rhnCapability.update_client_capabilities(self.server_id)
        # Fetch the channels this client is subscribed to
        channels = rhnChannel.getSubscribedChannels(self.server_id)

        # pylint: disable-next=invalid-name
        rhnServerTime = str(time.time())
        # pylint: disable-next=invalid-name
        expireOffset = str(CFG.CLIENT_AUTH_TIMEOUT)
        signature = computeSignature(
            CFG.SECRET_KEY, self.server_id, self.user, rhnServerTime, expireOffset
        )

        # pylint: disable-next=invalid-name
        loginDict = {
            "X-RHN-Server-Id": self.server_id,
            "X-RHN-Auth-User-Id": self.user,
            "X-RHN-Auth": signature,
            "X-RHN-Auth-Server-Time": rhnServerTime,
            "X-RHN-Auth-Expire-Offset": expireOffset,
            # List of lists [[label,last_mod],...]:
            "X-RHN-Auth-Channels": channels,
        }

        # Duplicate these values in the headers so that the proxy can
        # intercept and cache them without parseing the xmlrpc.
        transport = rhnFlags.get("outputTransportOptions")
        for k, v in list(loginDict.items()):
            # Special case for channels
            if k.lower() == "X-RHN-Auth-Channels".lower():
                # Concatenate the channel information column-separated
                transport[k] = [":".join(x) for x in v]
            else:
                transport[k] = v
        log_debug(5, "loginDict", loginDict, transport)

        # store route in DB (schema for RHN 3.1+ only!)
        server_route.store_client_route(self.server_id)

        return loginDict

    # pylint: disable-next=invalid-name
    def listChannels(self, system_id):
        """Clients v2+"""
        log_debug(5, system_id)
        # Authenticate the system certificate
        self.auth_system("listChannels", system_id)
        # log the entry
        log_debug(1, self.server_id)
        # pylint: disable-next=invalid-name
        channelList = rhnChannel.channels_for_server(self.server_id)
        return channelList

    # pylint: disable-next=invalid-name
    def subscribeChannels(self, system_id, channelNames, username, passwd):
        """Clients v2+"""
        add_to_seclist(passwd)
        log_debug(5, system_id, channelNames, username, passwd)
        # Authenticate the system certificate
        self.auth_system("subscribeChannel", system_id)
        # log the entry
        log_debug(1, self.server_id, channelNames)
        server_lib.snapshot_server(self.server_id, "Base Channel Updated")
        # pylint: disable-next=invalid-name
        for channelName in channelNames:
            if NONSUBSCRIBABLE_CHANNELS.search(channelName):
                raise rhnFault(73, explain=False)
            else:
                rhnChannel.subscribe_channel(
                    self.server_id, channelName, username, passwd
                )
        return 0

    # pylint: disable-next=invalid-name
    def unsubscribeChannels(self, system_id, channelNames, username, passwd):
        """Clients v2+"""
        add_to_seclist(passwd)
        log_debug(3)
        # Authenticate the system certificate
        self.auth_system("unsubscribeChannel", system_id)
        # log the entry
        log_debug(1, self.server_id, channelNames)
        # pylint: disable-next=invalid-name
        for channelName in channelNames:
            rhnChannel.unsubscribe_channel(
                self.server_id, channelName, username, passwd
            )
        return 0

    def solvedep(self, system_id, deps):
        """Clients v1-
        Solve dependencies for a given dependency problem list.
        IN:  a dependency problem list: [name, name, name, ...]
        RET: a package list: [[n,v,r,e],[n,v,r,e],...] That solves the
             dependencies.
        """
        log_debug(4, system_id)
        return self.__solveDep(system_id, deps, action="solvedep", clientVersion=1)

    # pylint: disable-next=invalid-name
    def solveDependencies(self, system_id, deps):
        """Clients v2+
        Solve dependencies for a given dependency problem list (newer version)
        IN:  a dependency problem list: [name, name, name, ...]
        RET: a hash {name: [[n, v, r, e], [n, v, r, e], ...], ...}
        """
        log_debug(4, system_id)
        return self.__solveDep(system_id, deps, action="solvedep", clientVersion=2)

    # pylint: disable-next=invalid-name
    def solveDependencies_arch(self, system_id, deps):
        """Does the same thing as solve_dependencies, but also returns the architecture label with the
        package info.
        IN:  a dependency problem list: [name, name, name, ...]
        RET: a hash {name: [[n, v, r, e, a], [n, v, r, e, a], ...], ...}
        """
        log_debug(4, system_id)
        return self.__solveDep_arch(system_id, deps, action="solvedep", clientVersion=2)

    # pylint: disable-next=invalid-name
    def solveDependencies_with_limits(
        self,
        system_id,
        deps,
        # pylint: disable-next=redefined-builtin
        all=0,
        limit_operator=None,
        limit=None,
    ):
        """This version of solve_dependencies allows the caller to get all of the packages that solve a
        dependency and limit the packages that are returned to those that match the criteria defined
        by limit_operator and limit. This version of the function also returns the architecture label
        of the package[s] that get returned.

        limit_operator can be any of: '<', '<=', '==', '>=', or '>'.
        limit is a a string of the format [epoch:]name-version-release
        deps is a list of filenames that the packages that are returned must provide.
        version is the version of the client that is calling the function.
        """
        log_debug(4, system_id)
        return self.__solveDep_with_limits(
            system_id,
            deps,
            action="solvedep",
            clientVersion=2,
            all=all,
            limit_operator=limit_operator,
            limit=limit,
        )

    def history(self, system_id, summary, body=""):
        """Clients v2+
        Add a history log for a performed action
        """
        log_debug(5, system_id, summary, body)
        # Authenticate the system certificate
        server = self.auth_system("history", system_id)
        # log the entry
        log_debug(1, self.server_id)
        # XXX: Probably this should be a non fatal error...
        server.add_history(summary, body)
        server.save_history()
        return 0

    # --- PRIVATE METHODS ---

    # pylint: disable-next=invalid-name
    def __solveDep_prepare(self, system_id, deps, action, clientVersion):
        """Response for clients:
        version 1: list
        version 2: hash
        """
        log_debug(7, system_id, deps, action, clientVersion)
        # pylint: disable-next=invalid-name
        faultString = _("Invalid value %s (%s)")
        if type(deps) not in (ListType, TupleType):
            log_error("Invalid argument type", type(deps))
            raise rhnFault(30, faultString % (deps, type(deps)))
        for dep in deps:
            # pylint: disable-next=unidiomatic-typecheck
            if type(dep) is not StringType:
                log_error("Invalid dependency member", type(dep))
                raise rhnFault(30, faultString % (dep, type(dep)))
        # Ignore empty strings
        deps = list(filter(len, deps))
        # anything left to do?
        if not deps:
            return []
        # Authenticate the system certificate
        # pylint: disable-next=unused-variable
        server = self.auth_system(action, system_id)
        # pylint: disable-next=consider-using-f-string
        log_debug(1, self.server_id, action, "items: %d" % len(deps))
        return deps

    # pylint: disable-next=invalid-name
    def __solveDep(self, system_id, deps, action, clientVersion):
        """Response for clients:
        version 1: list
        version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
            # Solve dependencies
            result = rhnDependency.solve_dependencies(
                self.server_id, result, clientVersion
            )
        return result

    # pylint: disable-next=invalid-name
    def __solveDep_arch(self, system_id, deps, action, clientVersion):
        """Response for clients:
        version 1: list
        version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
            # Solve dependencies
            result = rhnDependency.solve_dependencies_arch(
                self.server_id, result, clientVersion
            )
        return result

    # pylint: disable-next=invalid-name
    def __solveDep_with_limits(
        self,
        system_id,
        deps,
        action,
        # pylint: disable-next=invalid-name
        clientVersion,
        # pylint: disable-next=redefined-builtin
        all=0,
        limit_operator=None,
        limit=None,
    ):
        """Response for clients:
        version 1: list
        version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
            # Solve dependencies
            result = rhnDependency.solve_dependencies_with_limits(
                self.server_id, deps, clientVersion, all, limit_operator, limit
            )
        return result


class Servers(rhnHandler):

    """A class to handle the site selection..."""

    def __init__(self):
        """Servers Class Constructor."""
        rhnHandler.__init__(self)
        self.functions.append("get")
        self.functions.append("list")

    # pylint: disable-next=unused-argument
    def get(self, *junk):
        """Older funtion that can be a noop."""
        return []

    # pylint: disable-next=unused-argument
    def list(self, systemid=None):
        """Returns a list of available servers the client can connect to."""
        servers_list = [
            {
                "server": "xmlrpc.rhn.redhat.com",
                "handler": "/XMLRPC",
                "description": "XML-RPC Server",
                "location": "United States",
            },
        ]
        return servers_list
